src/cli/repl/commands/lineage.ts

```diff
@@ -0,0 +1,77 @@
+import type { ReplCommand } from './main'
+import { PipelineExecutor } from '../../../core/pipeline-executor'
+import { DEFAULT_DATAFLOW_PIPELINE } from '../../../core/steps/pipeline/default-pipelines'
+import type { RShell } from '../../../r-bridge/shell'
+import { requestFromInput } from '../../../r-bridge/retriever'
+import type { SingleSlicingCriterion } from '../../../slicing/criterion/parse'
+import { slicingCriterionToId } from '../../../slicing/criterion/parse'
+import type { NodeId } from '../../../r-bridge/lang-4.x/ast/model/processing/node-id'
+import type { OutgoingEdges } from '../../../dataflow/graph/graph'
+import type { DataflowGraphEdge } from '../../../dataflow/graph/edge'
+import { edgeIncludesType, EdgeType } from '../../../dataflow/graph/edge'
+import type { DataflowGraphVertexInfo } from '../../../dataflow/graph/vertex'
+import type { DataflowInformation } from '../../../dataflow/info'
+import type { NormalizedAst } from '../../../r-bridge/lang-4.x/ast/model/processing/decorate'
+
+function splitAt(str: string, idx: number) {
+	return [str.slice(0, idx), str.slice(idx)]
+}
+
+async function getDfg(shell: RShell, remainingLine: string) {
+	return await new PipelineExecutor(DEFAULT_DATAFLOW_PIPELINE, {
+		shell,
+		request: requestFromInput(remainingLine.trim())
+	}).allRemainingSteps()
+}
+
+function filterRelevantEdges(edge: DataflowGraphEdge) {
+	return edgeIncludesType(EdgeType.DefinedBy | EdgeType.DefinedByOnCall | EdgeType.Returns | EdgeType.Reads, edge.types)
+}
+
+function pushRelevantEdges(queue: [NodeId, DataflowGraphEdge][], outgoingEdges: OutgoingEdges) {
+	queue.push(...[...outgoingEdges].filter(([_, edge]) => filterRelevantEdges(edge)))
+	return queue
+}
+
+/**
+ * Get the lineage of a node in the dataflow graph
+ *
+ * @param criterion - The criterion to get the lineage of
+ * @param ast       - The normalized AST
+ * @param dfg       - The dataflow graph
+ * @returns The lineage of the node represented as a list of node ids
+ */
+export function getLineage(criterion: SingleSlicingCriterion, ast: NormalizedAst, dfg: DataflowInformation) {
+	const [vertex, outgoingEdges] = dfg.graph.get(slicingCriterionToId(criterion, ast)) as [DataflowGraphVertexInfo, OutgoingEdges]
```

Comment:
Is this really to silence the undefined case? :D

Reply:
Yep, 'cause `slicingCriterionToId` will throw on invalid criteria. So I didn't bother to add an additional guard. But I can change this if you insist.

Reply:
I had several problems with `as` silencing things recently so i would prefer an extra guard. But do as you think its better!